x-pack/plugins/actions/server/builtin_action_types/cases_webhook/service.ts

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import axios, { AxiosResponse } from 'axios';
+
+import { Logger } from '@kbn/core/server';
+import { isString } from 'lodash';
+import { renderMustacheStringNoEscape } from '../../lib/mustache_renderer';
+import {
+  createServiceError,
+  getObjectValueByKey,
+  getPushedDate,
+  removeSlash,
+  throwIfResponseIsNotValidSpecial,
+} from './utils';
+import {
+  CreateIncidentParams,
+  ExternalServiceCredentials,
+  ExternalService,
+  CasesWebhookPublicConfigurationType,
+  CasesWebhookSecretConfigurationType,
+  ExternalServiceIncidentResponse,
+  GetIncidentResponse,
+  UpdateIncidentParams,
+  CreateCommentParams,
+} from './types';
+
+import * as i18n from './translations';
+import { request } from '../lib/axios_utils';
+import { ActionsConfigurationUtilities } from '../../actions_config';
+
+export const createExternalService = (
+  actionId: string,
+  { config, secrets }: ExternalServiceCredentials,
+  logger: Logger,
+  configurationUtilities: ActionsConfigurationUtilities
+): ExternalService => {
+  const {
+    createCommentJson,
+    createCommentMethod,
+    createCommentUrl,
+    createIncidentJson,
+    createIncidentMethod,
+    createIncidentResponseKey,
+    createIncidentUrl: createIncidentUrlConfig,
+    getIncidentResponseCreatedDateKey,
+    getIncidentResponseExternalTitleKey,
+    getIncidentResponseUpdatedDateKey,
+    getIncidentUrl,
+    hasAuth,
+    headers,
+    incidentViewUrl,
+    updateIncidentJson,
+    updateIncidentMethod,
+    updateIncidentUrl,
+  } = config as CasesWebhookPublicConfigurationType;
+  const { password, user } = secrets as CasesWebhookSecretConfigurationType;
+  if (
+    !getIncidentUrl ||
+    !createIncidentUrlConfig ||
+    !incidentViewUrl ||
+    !updateIncidentUrl ||
+    (hasAuth && (!password || !user))
+  ) {
+    throw Error(`[Action]${i18n.NAME}: Wrong configuration.`);
+  }
+
+  const createIncidentUrl = removeSlash(createIncidentUrlConfig);
+
+  const axiosInstance = axios.create({
+    ...(hasAuth && isString(secrets.user) && isString(secrets.password)
+      ? { auth: { username: secrets.user, password: secrets.password } }
+      : {}),
+    ...(headers != null ? { headers } : {}),
```

Comment:
The data I am sending is as a string. To work with Jira/ServiceNow I need to add the header `content-type: application/json`. Do you (reviewer) think this should be a default (removable) or should we rely on the user to input the header? I followed the current webhook which relies on the user to input

Reply:
I think it is better to default it on the backend and document it. We are sending a JSON. Even if the data are as a string the server on the other end will parse them as JSON.